File: selenium-agent-mcp-starter/pages/checkout.page.ts
import { By } from 'selenium-webdriver';
import { BasePage } from './base.page';

export class CheckoutPage extends BasePage {
  private title = By.css('.title');
  private firstNameInput = By.id('first-name');
  private lastNameInput = By.id('last-name');
  private postalCodeInput = By.id('postal-code');
  private continueButton = By.id('continue');
  private finishButton = By.id('finish');
  private cancelButton = By.id('cancel');
  private errorMessage = By.css('[data-test="error"]');
  private completeHeader = By.css('.complete-header');
  private summaryTotal = By.css('.summary_total_label');

  async getTitleText(): Promise<string> {
    return this.getText(this.title);
  }

  async continueWithCustomer(firstName: string, lastName: string, postalCode: string): Promise<void> {
    await this.type(this.firstNameInput, firstName);
    await this.type(this.lastNameInput, lastName);
    await this.type(this.postalCodeInput, postalCode);
    await this.click(this.continueButton);
    await this.waitForUrlContains('checkout-step-two.html', 10000);
  }

  async continueEmpty(): Promise<void> {
    await this.jsClick(this.continueButton);
  }

  async getErrorMessage(): Promise<string> {
    return this.getText(this.errorMessage);
  }

  async finishCheckout(): Promise<void> {
    await this.jsClick(this.finishButton);
    await this.waitForUrlContains('checkout-complete.html');
  }

  async cancelCheckout(): Promise<void> {
    await this.jsClick(this.cancelButton);
    await this.waitForUrlContains('cart.html');
  }

  async getSummaryTotalText(): Promise<string> {
    return this.getText(this.summaryTotal);
  }

  async getCompleteHeaderText(): Promise<string> {
    return this.getText(this.completeHeader);
  }
}

File: selenium-agent-mcp-starter-python/src/pages/checkout_page.py
from selenium.webdriver.common.by import By
from src.pages.base_page import BasePage


class CheckoutPage(BasePage):
    _title = (By.CSS_SELECTOR, ".title")
    _first_name_input = (By.ID, "first-name")
    _last_name_input = (By.ID, "last-name")
    _postal_code_input = (By.ID, "postal-code")
    _continue_button = (By.ID, "continue")
    _finish_button = (By.ID, "finish")
    _error_message = (By.CSS_SELECTOR, "[data-test='error'], .error-message-container h3")
    _summary_container = (By.ID, "checkout_summary_container")
    _complete_header = (By.CSS_SELECTOR, ".complete-header, [data-test='complete-header']")
    _back_home_button = (By.ID, "back-to-products")

    def _set_value(self, locator: tuple, value: str):
        element = self.wait_for_visible_element(locator)
        self.driver.execute_script(
            """
            const element = arguments[0];
            const value = arguments[1];
            element.focus();
            element.value = value;
            element.dispatchEvent(new Event('input', { bubbles: true }));
            element.dispatchEvent(new Event('change', { bubbles: true }));
            """,
            element,
            value,
        )

    def get_title_text(self) -> str:
        return self.get_text(self._title)

    def fill_customer_info(self, first_name: str, last_name: str, postal_code: str):
        self._set_value(self._first_name_input, first_name)
        self._set_value(self._last_name_input, last_name)
        self._set_value(self._postal_code_input, postal_code)

    def continue_checkout(self):
        self.js_click(self._continue_button)
        if "checkout-step-one.html" in self.current_url():
            return
        self.wait_for_url_contains("checkout-step-two.html")

    def finish_checkout(self):
        self.js_click(self._finish_button)
        self.wait_for_url_contains("checkout-complete.html")

    def get_error_message(self) -> str:
        return self.get_text(self._error_message)

    def has_summary(self) -> bool:
        return self.is_visible(self._summary_container)

    def get_complete_header(self) -> str:
        return self.get_text(self._complete_header)

    def back_home(self):
        self.js_click(self._back_home_button)
        self.wait_for_url_contains("inventory.html")
